salt/orch/removal.sls

```diff
@@ -0,0 +1,394 @@
+    {%- set bootstrapped_etcd_members = salt.saltutil.runner('mine.get', tgt='G@roles:etcd and G@bootstrap_complete:true', fun='network.interfaces', tgt_type='compound').keys() %}
+    {%- if replacement in bootstrapped_etcd_members %}
+      {%- do salt.caasp_log.warn('CaaS: the replacement for the etcd server %s cannot be %s: another etcd server is already running there', target, replacement) %}
+      {%- if replacement_provided %}
+        {%- do salt.caasp_log.abort('CaaS: fatal!! the user provided replacement %s cannot be used', replacement) %}
+      {%- endif %}
+      {%- set replacement = '' %}
+    {%- endif %}
+  {%- endif %}
+
+  {%- if replacement %}
+    {%- do salt.caasp_log.debug('CaaS: setting %s as the replacement for the etcd member %s', replacement, target) %}
+    {%- do replacement_roles.append('etcd') %}
+  {%- elif etcd_members|length > 1 %}
+    {%- do salt.caasp_log.warn('CaaS: numnber of etcd members will be reduced to %d, as no replacement for %s has been found (or provided)', etcd_members|length, target) %}
+  {%- else %}
+    {%- do salt.caasp_log.abort('CaaS: cannot remove etcd member %s: too few etcd members, and no replacement found or provided', target) %}
+  {%- endif %}
+{%- endif %} {# target in etcd_members #}
+
+{#
+ # replacement for k8s masters
+ #}
+{%- if target in masters %}
+  {%- if not replacement %}
+    {# TODO: implement a replacement finder for k8s masters #}
+    {# NOTE: even if no `replacement` was provided in the pillar,
+     #       we probably have one at this point: if the master was
+     #       running etcd as well, we have already tried to find
+     #       a replacement in the previous step...
+     #       however, we must verify that the etcd replacement
+     #       is a valid k8s master replacement too.
+     #       (ideally we should find the union of etcd and
+     #       masters candidates)
+     #}
+  {%- endif %}
+
+  {# check if the replacement provided/found is valid #}
+  {%- if replacement %}
+    {%- set bootstrapped_masters = salt.saltutil.runner('mine.get', tgt='G@roles:kube-master and G@bootstrap_complete:true', fun='network.interfaces', tgt_type='compound').keys() %}
+    {%- if replacement in bootstrapped_masters %}
+      {%- do salt.caasp_log.warn('CaaS: error!! the replacement for an k8s master %s cannot be %s: another k8s master is already running there', target, replacement) %}
+      {%- if replacement_provided %}
+        {%- do salt.caasp_log.abort('CaaS: fatal!! the user provided replacement %s cannot be used', replacement) %}
+      {%- endif %}
+      {%- set replacement = '' %}
+    {%- elif replacement in minions %}
+      {%- do salt.caasp_log.warn('CaaS: warning!! will not replace the k8s master at %s: the replacement found/provided is the k8s minion %s', target, replacement) %}
+      {%- if replacement_provided %}
+        {%- do salt.caasp_log.abort('CaaS: fatal!! the user provided replacement %s cannot be used', replacement) %}
```

Comment:
Do we want to allow to provide a replacement? I think for a first implementation it would make sense to avoid this, and stick to the easiest implementation possible, as this kind of edge cases add unnecessary complexity IMO, while nothing will call to the orchestration with a replacement (at least Velum won't).

Reply:
We discussed in the planning meeting about this: a replacement could be provided, specially for masters.